base/utferror.jl

```diff
@@ -0,0 +1,51 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+#=
+@doc """
+@brief      Error messages for Unicode / UTF support
+""" ->
+=#
+
+const UTF_ERR_SHORT = 1
+const UTF_ERR_CONT  = 2
+const UTF_ERR_LONG  = 3
+const UTF_ERR_NOT_LEAD = 4
+const UTF_ERR_NOT_TRAIL = 5
+const UTF_ERR_NOT_SURROGATE = 6
+const UTF_ERR_MISSING_SURROGATE = 7
+const UTF_ERR_INVALID = 8
+const UTF_ERR_SURROGATE = 9
+const UTF_ERR_NULL_16_TERMINATE = 10
+const UTF_ERR_NULL_32_TERMINATE = 11
+const UTF_ERR_MAX = 11
+
+const errMsgs = [
+    "invalid UTF-8 sequence starting at index <<1>> (0x<<2>>) missing one or more continuation bytes)",
+    "invalid UTF-8 sequence starting at index <<1>> (0x<<2>> is not a continuation byte)",
+    "invalid UTF-8 sequence, overlong encoding starting at index <<1>> (0x<<2>>)",
+    "not a leading Unicode surrogate code unit at index <<1>> (0x<<2>>)",
+    "not a trailing Unicode surrogate code unit at index <<1>> (0x<<2>>)",
+    "not a valid Unicode surrogate code unit at index <<1>> (0x<<2>>",
+    "missing trailing Unicode surrogate code unit after index <<1>> (0x<<2>>)",
+    "invalid Unicode character starting at index <<1>> (0x<<2>> > 0x10ffff)",
+    "surrogate encoding not allowed in UTF-8 or UTF-32, at index <<1>> (0x<<2>>)",
+    "UTF16String data must be NULL-terminated",
+    "UTF32String data must be NULL-terminated"
+]
+#=
+@doc """
+@brief      Throws ArgumentError with information about the specific error, location, and character
+
+@param[in]  errcode Error code for Unicode error (one of UTF_ERR_*)
+@param[in]  charpos Index of invalid byte or character
+@param[in]  invchar Invalid byte or character
+
+@throws never returns, always throws ArgumentError
+""" ->
+=#
+function utf_errfunc(errcode::Integer, charpos, invchar)
```

Comment:
1) I actually wasn't trying to "argue from authority" at all.  @jakebolewski had talked about my motivations for making the change this way, and I was trying to answer that as best I could.
Are Julians *always* so sensitive about talking about one's experience, in the context of a technical discussion?  IMO, it can help shed light on a issue... I, for example, am very interested to learn about Julian's experiences in the technical/numeric/scientific computing world.  The type of arguing that I've seen here that does bother me is "argument from a *position* of authority", and people willing to use that position to curtail or stifle discussion.

2) I have tried to give the numbers and executive summaries as I went along, but with 500+ comments, I think the early ones where I tested different error handling strategies got lost in the chaff.  As soon as I have a minute (I *am* also busy working here in Belgium), I'll try to recreate my tests that pointed me to the issue of GC frame creation [and I know that @yuyichao is working hard to fix things so it wouldn't even be an issue].  I *did* give an executive summary just before I went to bed, did you miss it?  21-100% slower without the `@inbounds` macros... (I think more frequently closer to 100%).

3) Well, first off, I haven't really seen that much consistency in the Julia code I've been looking at...
(maybe because it is older code, who knows), so I don't totally buy the argument that *everything* must be written in "idiomatic" Julia code.  Different people also have different ideas of what constitutes good structure and practice...  Does Julia really need some sort of unwritten style that everybody must follow? (at least, if it were *written down*, people might have a chance of knowing what was expected, and using that, or debating whether the current practice really is good...)
I'm not sure at all that the suggestions that I haven't incorporated so far would make things "substantially better"...

About the error handling: the reason the *very first thing* that I implemented was the better error handling, is that I would *not* have been able to easily find and fix all of the many bugs in the current code without the enhanced messages.   It all ties together...  this PR adds some extra overhead already to do complete validity checking, which people worried about in the initial comments on #11004... A big goal of mine was to at the very least, not introduce any regressions in performance (the performance was so bad already, that would have been criminal).  Because of my approaches, not only is the code not slower, even with the extra checking, it is faster as well.
At least in my early benchmarking, the extra overhead in the inner loops (which is *where* the error checking happens!) made the performance frequently worse than the current bad performance.

All of the above is why I really do not think it is at all a good think to try to split this up as has been suggested.